python/dgl/contrib/dis_kvstore.py

```diff
@@ -679,55 +701,57 @@ def pull(self, name, id_tensor):
         assert len(name) > 0, 'name cannot be empty.'
         assert F.ndim(id_tensor) == 1, 'ID must be a vector.'
 
-        if len(self._garbage_msg) > 1000:
+        if len(self._garbage_msg) > GARBAGE_COLLECTION_COUNT:
             _clear_kv_msg(self._garbage_msg)
             self._garbage_msg = []
 
-        # partition data (we can move this part of code into C-api if needed)
-        server_id = self._data_store[name+'-part-'][id_tensor]
-        # sort index by server id
-        sorted_id = np.argsort(F.asnumpy(server_id))
-        # we need return data with original order of ID
-        back_sorted_id = F.tensor(np.argsort(sorted_id))
-        id_tensor = id_tensor[F.tensor(sorted_id)]
-        server, count = np.unique(F.asnumpy(server_id), return_counts=True)
-        # pull data from server by server order
+        # partition data
+        machine_id = self._data_store[name+'-part-'][id_tensor]
+        # sort index by machine id
+        sorted_id = F.tensor(np.argsort(F.asnumpy(machine_id)))
+        back_sorted_id = F.tensor(np.argsort(F.asnumpy(sorted_id)))
+        id_tensor = id_tensor[sorted_id]
+        machine, count = np.unique(F.asnumpy(machine_id), return_counts=True)
+        # pull data from server by order
         start = 0
         pull_count = 0
-        local_data = {}
-        for idx in range(len(server)):
+        local_id = None
+
+        for idx in range(len(machine)):
             end = start + count[idx]
-            if start == end:  # don't have any data in target server
+            if start == end: # No data for target machine
                 continue
             partial_id = id_tensor[start:end]
-
-            if server[idx] in self._local_server_id and self._close_shared_mem == False:
-                if (name+'-g2l-'+str(server[idx]) in self._has_data) == True:
-                    local_id = self._data_store[name+'-g2l-'+str(server[idx])][partial_id]
+            if machine[idx] == self._machine_id: # local pull
+                if (name+'-g2l-' in self._has_data) == True:
+                    local_id = self._data_store[name+'-g2l-'][partial_id]
                 else:
                     local_id = partial_id
-                local_data[server[idx]] = self._pull_handler(name+'-data-'+str(server[idx]), local_id, self._data_store)
-            else:
+            else: # pull data from remote server
                 msg = KVStoreMsg(
                     type=KVMsgType.PULL, 
                     rank=self._client_id, 
                     name=name, 
-                    id=partial_id, 
+                    id=partial_id,
                     data=None,
                     c_ptr=None)
-                _send_kv_msg(self._sender, msg, server[idx])
+                s_id = random.randint(machine[idx]*self._backup_count, (machine[idx]+1)*self._backup_count-1)
```

Comment:
Why we use random.randint here?

Reply:
We need to select a random server ID for communicating.